File: arhiv/src/dto.ts
import { JSONObj, NominalType, Obj } from 'utils';

export type APIRequest =
  | {
      typeName: 'ListDocuments';
      documentTypes: DocumentType[];
      query: string;
      page: number;
    }
  | {
      typeName: 'GetDocuments';
      ids: DocumentId[];
    }
  | {
      typeName: 'GetStatus';
    }
  | {
      typeName: 'GetDocument';
      id: DocumentId;
    }
  | {
      typeName: 'ParseMarkup';
      markup: string;
    }
  | {
      typeName: 'SaveDocument';
      id: DocumentId;
      subtype: DocumentSubtype;
      data: DocumentData;
      collections: DocumentId[];
    }
  | {
      typeName: 'CreateDocument';
      documentType: DocumentType;
      subtype: DocumentSubtype;
      data: DocumentData;
      collections: DocumentId[];
    }
  | {
      typeName: 'EraseDocument';
      id: DocumentId;
    }
  | {
      typeName: 'ListDir';
      dir?: string;
      showHidden: boolean;
    }
  | {
      typeName: 'CreateAttachment';
      filePath: string;
      moveFile: boolean;
    }
  | {
      typeName: 'UploadFile';
      base64Data: string;
      fileName: string;
    }
  | {
      typeName: 'Scrape';
      url: string;
    }
  | {
      typeName: 'CommitOrSync';
    };

export type APIResponse =
  | {
      typeName: 'ListDocuments';
      documents: ListDocumentsResult[];
      hasMore: boolean;
    }
  | {
      typeName: 'GetDocuments';
      documents: ListDocumentsResult[];
    }
  | {
      typeName: 'GetStatus';
      status: string;
    }
  | {
      typeName: 'GetDocument';
      id: DocumentId;
      documentType: DocumentType;
      title: string;
      subtype: DocumentSubtype;
      updatedAt: string;
      data: DocumentData;
      backrefs: DocumentBackref[];
      collections: DocumentBackref[];
    }
  | {
      typeName: 'ParseMarkup';
      ast: MarkupElement;
    }
  | {
      typeName: 'SaveDocument';
      errors?: SaveDocumentErrors;
    }
  | {
      typeName: 'CreateDocument';
      id?: DocumentId;
      errors?: SaveDocumentErrors;
    }
  | {
      typeName: 'EraseDocument';
    }
  | {
      typeName: 'ListDir';
      dir: string;
      entries: DirEntry[];
    }
  | {
      typeName: 'CreateAttachment';
      id: DocumentId;
    }
  | {
      typeName: 'UploadFile';
      id: DocumentId;
    }
  | {
      typeName: 'Scrape';
      documents: ListDocumentsResult[];
    }
  | {
      typeName: 'CommitOrSync';
    };

export type DocumentId = NominalType<string, 'DocumentId'>;
export type DocumentType = NominalType<string, 'DocumentType'>;
export type DocumentSubtype = NominalType<string, 'DocumentSubtype'>;

export type DocumentDTO = Omit<Extract<APIResponse, { typeName: 'GetDocument' }>, 'typeName'>;

export const ERASED_DOCUMENT_TYPE = '' as DocumentType;
export const ATTACHMENT_DOCUMENT_TYPE = 'attachment' as DocumentType;
export const DEFAULT_SUBTYPE = '' as DocumentSubtype;
export const EMPTY_DATA: DocumentData = {};

export type ListDocumentsResult = {
  id: DocumentId;
  documentType: DocumentType;
  subtype: DocumentSubtype;
  title: string;
  updatedAt: string;
};

export type DocumentData = JSONObj;

export type DocumentBackref = {
  id: DocumentId;
  documentType: DocumentType;
  subtype: DocumentSubtype;
  title: string;
};

export type DocumentFieldErrors = Obj<string[]>;

export type SaveDocumentErrors = {
  documentErrors: string[];
  fieldErrors: DocumentFieldErrors;
};

export type DirEntry =
  | {
      typeName: 'Dir';
      name: string;
      path: string;
      isReadable: boolean;
    }
  | {
      typeName: 'File';
      name: string;
      path: string;
      isReadable: boolean;
      size: number;
    }
  | {
      typeName: 'Symlink';
      name: string;
      path: string;
      isReadable: boolean;
      linksTo: string;
      size?: number;
    };

export type FileEntry = Extract<DirEntry, { typeName: 'File' }>;

export type MarkupElement =
  | {
      typeName: 'Document';
      children: MarkupElement[];
    }
  | {
      typeName: 'Text';
      range: Range;
      value: string;
    }
  | {
      typeName: 'Code';
      range: Range;
      value: string;
    }
  | {
      typeName: 'Html';
      range: Range;
      value: string;
    }
  | {
      typeName: 'FootnoteReference';
      range: Range;
      label: string;
    }
  | {
      typeName: 'SoftBreak';
      range: Range;
    }
  | {
      typeName: 'HardBreak';
      range: Range;
    }
  | {
      typeName: 'Rule';
      range: Range;
    }
  | {
      typeName: 'TaskListMarker';
      range: Range;
      checked: boolean;
    }
  | {
      typeName: 'Paragraph';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Heading';
      range: Range;
      level: 'H1' | 'H2' | 'H3' | 'H4' | 'H5' | 'H6';
      children: MarkupElement[];
    }
  | {
      typeName: 'BlockQuote';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'CodeBlock';
      range: Range;
      kind: 'Indented' | { Fenced: string };
      children: MarkupElement[];
    }
  | {
      typeName: 'List';
      range: Range;
      first_item_number?: number;
      children: MarkupElement[];
    }
  | {
      typeName: 'ListItem';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'FootnoteDefinition';
      range: Range;
      label: string;
      children: MarkupElement[];
    }
  | {
      typeName: 'Table';
      range: Range;
      alignments: Array<'None' | 'Left' | 'Center' | 'Right'>;
      children: MarkupElement[];
    }
  | {
      typeName: 'TableHead';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'TableRow';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'TableCell';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Emphasis';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Strong';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Strikethrough';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Link';
      range: Range;
      link_type: LinkType;
      url: string;
      children: MarkupElement[];
    }
  | {
      typeName: 'Image';
      range: Range;
      link_type: LinkType;
      url: string;
      children: MarkupElement[];
    };

export type Range = {
  start: number;
  end: number;
};

type LinkType =
  | 'Inline'
  | 'Reference'
  | 'ReferenceUnknown'
  | 'Collapsed'
  | 'CollapsedUnknown'
  | 'Shortcut'
  | 'ShortcutUnknown'
  | 'Autolink'
  | 'Email';

export function throwBadMarkupElement(value: never): never;
export function throwBadMarkupElement(value: MarkupElement) {
  throw new Error(`Unknown MarkupElement: ${value.typeName}`);
}

File: arhiv/src/dto.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use baza::entities::{BLOBId, DocumentData, DocumentLockKey, Id};
use rs_utils::Timestamp;

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum APIRequest {
    #[serde(rename_all = "camelCase")]
    ListDocuments {
        document_types: Vec<String>,
        query: String,
        page: u8,
    },
    GetDocuments {
        ids: Vec<Id>,
    },
    GetStatus {},
    GetDocument {
        id: Id,
    },
    ParseMarkup {
        markup: String,
    },
    #[serde(rename_all = "camelCase")]
    CreateDocument {
        document_type: String,
        data: DocumentData,
        collections: Vec<Id>,
    },
    #[serde(rename_all = "camelCase")]
    SaveDocument {
        lock_key: DocumentLockKey,
        id: Id,
        data: DocumentData,
        collections: Vec<Id>,
    },
    EraseDocument {
        id: Id,
    },
    #[serde(rename_all = "camelCase")]
    ListDir {
        dir: Option<String>,
        show_hidden: bool,
    },
    #[serde(rename_all = "camelCase")]
    CreateAttachment {
        file_path: String,
        move_file: bool,
    },
    Commit {},
    Sync {},
    GetSaveState {},
    LockDocument {
        id: Id,
    },
    #[serde(rename_all = "camelCase")]
    UnlockDocument {
        id: Id,
        lock_key: Option<DocumentLockKey>,
        force_unlock: Option<bool>,
    },
    #[serde(rename_all = "camelCase")]
    ReorderCollectionRefs {
        collection_id: Id,
        id: Id,
        new_pos: usize,
    },
}

#[derive(Serialize)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum APIResponse {
    ListDocuments {
        documents: Vec<ListDocumentsResult>,
        #[serde(rename = "hasMore")]
        has_more: bool,
    },
    GetDocuments {
        documents: Vec<GetDocumentsResult>,
    },
    GetStatus {
        status: String,
    },
    #[serde(rename_all = "camelCase")]
    GetDocument {
        id: Id,
        title: String,
        document_type: String,
        updated_at: Timestamp,
        data: DocumentData,
        refs: Vec<Id>,
        backrefs: Vec<DocumentBackref>,
        collections: Vec<DocumentBackref>,
    },
    ParseMarkup {
        ast: Value,
    },
    CreateDocument {
        id: Option<Id>,
        errors: Option<SaveDocumentErrors>,
    },
    SaveDocument {
        errors: Option<SaveDocumentErrors>,
    },
    EraseDocument {},
    ListDir {
        dir: String,
        entries: Vec<DirEntry>,
    },
    CreateAttachment {
        id: Id,
    },
    Commit {},
    Sync {},
    #[serde(rename_all = "camelCase")]
    GetSaveState {
        can_commit: bool,
        can_sync: bool,
    },
    #[serde(rename_all = "camelCase")]
    LockDocument {
        lock_key: DocumentLockKey,
    },
    UnlockDocument {},
    ReorderCollectionRefs {},
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListDocumentsResult {
    pub id: Id,
    pub document_type: String,
    pub title: String,
    pub updated_at: Timestamp,
    pub data: DocumentData,
    pub cover: Option<BLOBId>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GetDocumentsResult {
    pub id: Id,
    pub document_type: String,
    pub title: String,
    pub updated_at: Timestamp,
    pub data: DocumentData,
}

#[derive(Serialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SaveDocumentErrors {
    pub document_errors: Vec<String>,
    pub field_errors: HashMap<String, Vec<String>>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentBackref {
    pub id: Id,
    pub document_type: String,
    pub title: String,
}

#[derive(Serialize)]
#[serde(tag = "typeName")]
pub enum DirEntry {
    #[serde(rename_all = "camelCase")]
    Dir {
        name: String,
        path: String,
        is_readable: bool,
    },

    #[serde(rename_all = "camelCase")]
    File {
        name: String,
        path: String,
        is_readable: bool,
        size: u64,
    },

    #[serde(rename_all = "camelCase")]
    Symlink {
        name: String,
        path: String,
        is_readable: bool,
        links_to: String,
        size: Option<u64>,
    },
}

impl DirEntry {
    pub fn get_name(&self) -> &str {
        match self {
            DirEntry::Dir { name, .. } => name,
            DirEntry::File { name, .. } => name,
            DirEntry::Symlink { name, .. } => name,
        }
    }
}
